functionary/ui/views/generic.py

```diff
@@ -0,0 +1,195 @@
+  object (PermissionedDetailView, PermissionedUpdateView, PermissionedDeleteView),
+  attempting to access the view for an object outside of the active environment will
+  result in a 404 response.
+
+- For views that are served via a nested route, the parent object is also used to filter
+  the queryset. To take advantage of this, the nested route must be defined with the
+  kwarg for the parent object named `<model>_pk`. For example, the path for a
+  WorkflowStep view, which has a parent object of a Workflow would be defined as:
+
+    workflow/<uuid:workflow_pk>/step/<uuid:pk>/edit
+
+  This would result in the base queryset looking something like:
+
+    WorkflowStep.objects.filter(workflow__pk=workflow_pk, pk=pk)
+
+Classes inheriting from any of these base views can optionally define the following
+properties:
+
+- environment_through_field: If the view's model does not itself have an environment
+    field, but is instead infers its environment through a related object, you can
+    declare that here, via a string that will be used as a Queryset filter kwarg. Using
+    WorkflowStep as an example again, we would set environment_through_field to
+    "workflow".
+- permissioned_model: This can be set so that the user's permission is checked against
+    a model other than the one defined by the view's model. This is useful if access to
+    an object is implied by access to some other model, generally a related model. For
+    example, access to view an EnvironmentUserRole may be implied if a user has access
+    to view the Environment. In this case, we can set permissioned_model to
+    "Environment." Note that the value is a string rather than an actual model class.
+    This is so that the model class isn't required to be imported just to use it as the
+    permissioned_model. Additionally, setting this value changes the required permission
+    from <MODEL>_<ACTION> to <MODEL>_UPDATE, as the operation being performed is treated
+    as an update to the permissioned model, rather than a create, update, or delete to
+    the view model. For example, deleting a WorkflowStep is treated as an UPDATE to the
+    Workflow.
+"""
+import re
+from functools import cache
+from typing import Type
+
+from django.contrib.auth.mixins import LoginRequiredMixin, UserPassesTestMixin
+from django.db import models
+from django.views.generic import (
+    CreateView,
+    DeleteView,
+    DetailView,
+    ListView,
+    UpdateView,
+)
+
+from core.auth import Permission
+from core.models import Environment
+
+
+class PermissionedViewMixin(LoginRequiredMixin, UserPassesTestMixin):
+    """Mixin for building generic (model based) views.
+
+    This provides general utilities for handling the session environment and checking
+    user permissions against the environment.
+    """
+
+    environment_through_field: str | None = None
+    model: Type[models.Model] | None = None
+    permissioned_model: str | None = None
```

Comment:
Might be worth including a brief docstring inside the class explaining what the `environment_through_field`, `model`, `post_action`, and `permissioned_model` variables do. This would allow others to just hover over the class in their editor to view the docstring about the view, versus having to go to the `generic.py` to view the class themselves

Reply:
It's already the case that you can hover over the module and see the full docstring with all of the details.  I really don't want to duplicate the info there across all of the classes because then there's 5 places to update docs when things change.  What I did instead is just add "See `ui.views.generic` for details" to each of the class docstrings so that at least if you hover over the class it informs you of where you should actually look for the info.